File: include/extras-kguiaddons/libkwordwrap.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api-staging.kde.org/kwordwrap.html
pub const kwordwrap = struct {
    /// New constructs a new KWordWrap object.
    ///
    /// ``` other: QtC.KWordWrap ```
    pub fn New(other: ?*anyopaque) QtC.KWordWrap {
        return qtc.KWordWrap_new(@ptrCast(other));
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#formatText)
    ///
    /// ``` fm: QtC.QFontMetrics, r: QtC.QRect, flags: i32, str: []const u8 ```
    pub fn FormatText(fm: ?*anyopaque, r: ?*anyopaque, flags: i32, str: []const u8) QtC.KWordWrap {
        const str_str = qtc.libqt_string{
            .len = str.len,
            .data = str.ptr,
        };
        return qtc.KWordWrap_FormatText(@ptrCast(fm), @ptrCast(r), @intCast(flags), str_str);
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#boundingRect)
    ///
    /// ``` self: QtC.KWordWrap ```
    pub fn BoundingRect(self: ?*anyopaque) QtC.QRect {
        return qtc.KWordWrap_BoundingRect(@ptrCast(self));
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#wrappedString)
    ///
    /// ``` self: QtC.KWordWrap, allocator: std.mem.Allocator ```
    pub fn WrappedString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KWordWrap_WrappedString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kwordwrap.WrappedString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#truncatedString)
    ///
    /// ``` self: QtC.KWordWrap, allocator: std.mem.Allocator ```
    pub fn TruncatedString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KWordWrap_TruncatedString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kwordwrap.TruncatedString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#drawText)
    ///
    /// ``` self: QtC.KWordWrap, painter: QtC.QPainter, x: i32, y: i32 ```
    pub fn DrawText(self: ?*anyopaque, painter: ?*anyopaque, x: i32, y: i32) void {
        qtc.KWordWrap_DrawText(@ptrCast(self), @ptrCast(painter), @intCast(x), @intCast(y));
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#operator-eq)
    ///
    /// ``` self: QtC.KWordWrap, other: QtC.KWordWrap ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.KWordWrap_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#drawFadeoutText)
    ///
    /// ``` p: QtC.QPainter, x: i32, y: i32, maxW: i32, t: []const u8 ```
    pub fn DrawFadeoutText(p: ?*anyopaque, x: i32, y: i32, maxW: i32, t: []const u8) void {
        const t_str = qtc.libqt_string{
            .len = t.len,
            .data = t.ptr,
        };
        qtc.KWordWrap_DrawFadeoutText(@ptrCast(p), @intCast(x), @intCast(y), @intCast(maxW), t_str);
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#drawTruncateText)
    ///
    /// ``` p: QtC.QPainter, x: i32, y: i32, maxW: i32, t: []const u8 ```
    pub fn DrawTruncateText(p: ?*anyopaque, x: i32, y: i32, maxW: i32, t: []const u8) void {
        const t_str = qtc.libqt_string{
            .len = t.len,
            .data = t.ptr,
        };
        qtc.KWordWrap_DrawTruncateText(@ptrCast(p), @intCast(x), @intCast(y), @intCast(maxW), t_str);
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#formatText)
    ///
    /// ``` fm: QtC.QFontMetrics, r: QtC.QRect, flags: i32, str: []const u8, lenVal: i32 ```
    pub fn FormatText5(fm: ?*anyopaque, r: ?*anyopaque, flags: i32, str: []const u8, lenVal: i32) QtC.KWordWrap {
        const str_str = qtc.libqt_string{
            .len = str.len,
            .data = str.ptr,
        };
        return qtc.KWordWrap_FormatText5(@ptrCast(fm), @ptrCast(r), @intCast(flags), str_str, @intCast(lenVal));
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#truncatedString)
    ///
    /// ``` self: QtC.KWordWrap, dots: bool, allocator: std.mem.Allocator ```
    pub fn TruncatedString1(self: ?*anyopaque, dots: bool, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KWordWrap_TruncatedString1(@ptrCast(self), dots);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kwordwrap.TruncatedString1: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#drawText)
    ///
    /// ``` self: QtC.KWordWrap, painter: QtC.QPainter, x: i32, y: i32, flags: i32 ```
    pub fn DrawText4(self: ?*anyopaque, painter: ?*anyopaque, x: i32, y: i32, flags: i32) void {
        qtc.KWordWrap_DrawText4(@ptrCast(self), @ptrCast(painter), @intCast(x), @intCast(y), @intCast(flags));
    }

    /// [Qt documentation](https://api-staging.kde.org/kwordwrap.html#dtor.KWordWrap)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KWordWrap ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KWordWrap_Delete(@ptrCast(self));
    }
};

/// https://api-staging.kde.org/kwordwrap.html#types
pub const enums = struct {
    pub const KWordWrap = enum {
        pub const FadeOut: i32 = 268435456;
        pub const Truncate: i32 = 536870912;
    };
};

File: src_original/extras-kguiaddons/libkwordwrap.cpp
#include <KWordWrap>
#include <QFontMetrics>
#include <QPainter>
#include <QRect>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <kwordwrap.h>
#include "libkwordwrap.h"
#include "libkwordwrap.hxx"

KWordWrap* KWordWrap_new(const KWordWrap* other) {
    return new KWordWrap(*other);
}

KWordWrap* KWordWrap_FormatText(QFontMetrics* fm, const QRect* r, int flags, const libqt_string str) {
    QString str_QString = QString::fromUtf8(str.data, str.len);
    return new KWordWrap(KWordWrap::formatText(*fm, *r, static_cast<int>(flags), str_QString));
}

QRect* KWordWrap_BoundingRect(const KWordWrap* self) {
    return new QRect(self->boundingRect());
}

libqt_string KWordWrap_WrappedString(const KWordWrap* self) {
    QString _ret = self->wrappedString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KWordWrap_TruncatedString(const KWordWrap* self) {
    QString _ret = self->truncatedString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void KWordWrap_DrawText(const KWordWrap* self, QPainter* painter, int x, int y) {
    self->drawText(painter, static_cast<int>(x), static_cast<int>(y));
}

void KWordWrap_OperatorAssign(KWordWrap* self, const KWordWrap* other) {
    self->operator=(*other);
}

void KWordWrap_DrawFadeoutText(QPainter* p, int x, int y, int maxW, const libqt_string t) {
    QString t_QString = QString::fromUtf8(t.data, t.len);
    KWordWrap::drawFadeoutText(p, static_cast<int>(x), static_cast<int>(y), static_cast<int>(maxW), t_QString);
}

void KWordWrap_DrawTruncateText(QPainter* p, int x, int y, int maxW, const libqt_string t) {
    QString t_QString = QString::fromUtf8(t.data, t.len);
    KWordWrap::drawTruncateText(p, static_cast<int>(x), static_cast<int>(y), static_cast<int>(maxW), t_QString);
}

KWordWrap* KWordWrap_FormatText5(QFontMetrics* fm, const QRect* r, int flags, const libqt_string str, int lenVal) {
    QString str_QString = QString::fromUtf8(str.data, str.len);
    return new KWordWrap(KWordWrap::formatText(*fm, *r, static_cast<int>(flags), str_QString, static_cast<int>(lenVal)));
}

libqt_string KWordWrap_TruncatedString1(const KWordWrap* self, bool dots) {
    QString _ret = self->truncatedString(dots);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void KWordWrap_DrawText4(const KWordWrap* self, QPainter* painter, int x, int y, int flags) {
    self->drawText(painter, static_cast<int>(x), static_cast<int>(y), static_cast<int>(flags));
}

void KWordWrap_Delete(KWordWrap* self) {
    delete self;
}
